### PR TITLE
Oidc Auth for Azure steps

### PR DESCRIPTION
**Pull request by IsaacCalligeros95**
This PR supports AzureOidc Accounts to Calamari for all Azure steps **excluding CloudService & ServiceFabric**. The associated Server changes can be found [here](https://github.com/OctopusDeploy/OctopusDeploy/pull/20042). 

As a part of this change, ServicePrincipal was implemented mostly per CalamariFlavour basis. This has been consolidated into the _Calamari.CloudAccounts_ repository, which is now shared across most Azure Calamari repositories. This adds some unnecessary overhead to the Calamari Bundles for some of these flavours, but the included Nuget Packages are relatively small ~1MB for the largest, and given the additional duplication introduced for AzureOidc Accounts looks to be worth the cost of consolidating.

[See Server PR for test cases](https://github.com/OctopusDeploy/OctopusDeploy/pull/20042).)
[sc-57210}

### CODE REVIEW

File: source/Calamari.Azure/ServicePrincipalAccount.cs
Comment (IsaacCalligeros95) [inline source/Calamari.Azure/ServicePrincipalAccount.cs:10]: There's no actual shared base project for Calamari Azure steps. This one is specifically for Kubernetes cloud target discovery and is only included in the base Calamari Project. I'd like to pull the ServicePrincipalAccount, AzureOidcAccount and ideally all accounts into Calamari.Common which is used in the majority of steps. I have not done this as a part of this PR, at least not yet any feedback or hesitations on making this change?

Comment (IsaacCalligeros95) [reply source/Calamari.Azure/ServicePrincipalAccount.cs:10]: An initial cut at this required bringing in the following packages to Calamari.Common. Alterntivly these could just be extension methods in the individual flavours that requite these (AzureWebApp & Calamari.Azure) It's likely worthwhile doing this but I'm holding off for feedback

`using Microsoft.Azure.Management.Fluent;
using Microsoft.Azure.Management.ResourceManager.Fluent.Authentication;
using Microsoft.Rest;`

File: source/Calamari.AzureAppService/Calamari.AzureAppService.csproj
Comment (IsaacCalligeros95) [inline source/Calamari.AzureAppService/Calamari.AzureAppService.csproj:27]: 3.1.1 returns a 404 on Access Tokens Generated through OAuth flows

File: source/Calamari.AzureAppService/Behaviors/TargetDiscoveryBehaviour.cs
Comment (IsaacCalligeros95) [inline source/Calamari.AzureAppService/Behaviors/TargetDiscoveryBehaviour.cs:114]: Oidc Accounts are not supported in Calamari yet

File: source/Calamari.CloudAccounts/Calamari.CloudAccounts.csproj
Comment (IsaacCalligeros95) [inline source/Calamari.CloudAccounts/Calamari.CloudAccounts.csproj:25]: Net452 limits to version 10.0.3, rather than the later 13.0.3 used in other projects

File: source/Calamari.AzureAppService/Behaviors/TargetDiscoveryBehaviour.cs
Comment (IsaacCalligeros95) [inline source/Calamari.AzureAppService/Behaviors/TargetDiscoveryBehaviour.cs:18]: Switched to Newtonsoft.Json deserialization as Calamari.CloudAccounts builds net452 and System.Text.Json is only available from net461 on. Newtonsoft is case insensitive by default which matches the existing behaviour

File: source/Calamari.AzureAppService/AppSettingsManagement.cs
Comment (benPearce1) [inline source/Calamari.AzureAppService/AppSettingsManagement.cs:101]: did this cause a problem when you were testing?

Comment (IsaacCalligeros95) [reply source/Calamari.AzureAppService/AppSettingsManagement.cs:101]: NewtonSoft.Json returns a nullable string compared to System.Text.Json returning a string. I had to swap to NewtonSoft because of the Calamari.CloudAccounts net452 build

File: source/Calamari.AzureAppService/Azure/AzureClient.cs
Comment (IsaacCalligeros95) [reply source/Calamari.AzureAppService/Azure/AzureClient.cs:31]: Pushed down to the Calamari.CloudAccounts, each of AzureServicePrincipal and AzureOidcAccount have their implementation of this method with a shared interface.

File: source/Calamari.AzureAppService/Calamari.AzureAppService.csproj
Comment (benPearce1) [inline source/Calamari.AzureAppService/Calamari.AzureAppService.csproj:33]: Was there a specific reason for swapping to NewtonSoft?
Sometimes, serialisation libraries can give you different results if you don't setup the configuration right

Comment (IsaacCalligeros95) [reply source/Calamari.AzureAppService/Calamari.AzureAppService.csproj:33]: Calamari.CloudAccounts builds net452 and System.Text.Json is only available from net461 on. Newtonsoft is case insensitive by default, which matches the existing behaviour we specified in System.Text.Json in the `source/Calamari.AzureAppService/Behaviors/TargetDiscoveryBehaviour.cs`

Comment (benPearce1) [reply source/Calamari.AzureAppService/Calamari.AzureAppService.csproj:33]: sorry, just saw your response from earlier

File: source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs
Comment (benPearce1) [inline source/Calamari.AzureAppService/Behaviors/AzureAppServiceBehaviour.cs:1]: empty file. should be removed, doesn't have to be this pr

File: source/Calamari.AzureScripting/AzureContextScriptWrapper.cs
Comment (benPearce1) [inline source/Calamari.AzureScripting/AzureContextScriptWrapper.cs:80]: The variables being set here are a bit confusing. For an OIDC account we get `OctopusAzSpOrOidc` set to `true`, but for a password based service principal account we get `OctopusAzSpOrOidc` and `OctopusUseServicePrincipal` set to `true`.
Don't think it is worth changing, but food for thought. Also, not a massive fan of the shortened `OctopusAzSpOrOidc`, would prefer to spell it all out, `OctopusAzureServicePrincipalOrOidc` 🤷

**Review by benPearce1 — APPROVED** [review]
This is good. There are some really minor points for improvement, but I would like to merge this and review later when we/if have time.

**Comment by IsaacCalligeros95** [comment]
> This is good. There are some really minor points for improvement, but I would like to merge this and review later when we/if have time.

Addressed comments in branch for Terraform [work here](https://github.com/OctopusDeploy/Calamari/commit/c20faf61088d50a9e61e786f798c0b565862ab79)

File: source/Calamari.AzureScripting/AzureContextScriptWrapper.cs
Comment (IsaacCalligeros95) [reply source/Calamari.AzureScripting/AzureContextScriptWrapper.cs:80]: I'd meant to tidy this up, I think the scripts mostly had this removed but missed the wrapper. Thanks for the call out

**Pull request merged**
